File: app/src/main/java/unforgettable/azcs/me/unforgettable/AddWordActivity.java
package unforgettable.azcs.me.unforgettable;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;

public class AddWordActivity extends AppCompatActivity {
    EditText word, meaning, sentence;
    RecyclerView sentencelist;
    Button addButton;
    FirebaseDatabase database;
    DatabaseReference myRef;
    DatabaseReference wordIdRef;
    DatabaseReference sentenceIdRef;
    FirebaseUser user;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_word);
        word = findViewById(R.id.word);
        meaning = findViewById(R.id.meaning);
        sentence = findViewById(R.id.sentence);
        sentencelist = findViewById(R.id.sentencesList);
        addButton = findViewById(R.id.addbtn);
        database = FirebaseDatabase.getInstance();
        user = FirebaseAuth.getInstance().getCurrentUser();
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (Utils.isValid(word) && Utils.isValid(meaning)) {
                    myRef = database.getReference(user.getUid());
                    wordIdRef = myRef.push();
                    myRef.child(wordIdRef.getKey()).setValue(getWord());
                    startActivity(new Intent(AddWordActivity.this, MainActivity.class));
                }
            }
        });


    }

    private Word getWord() {

        if (TextUtils.isEmpty(sentence.getText().toString()))
            return new Word(wordIdRef.getKey(), word.getText().toString(), meaning.getText().toString(), null);

        sentenceIdRef = wordIdRef.push();
        final Sentence s = new Sentence(sentence.getText().toString(), sentenceIdRef.getKey());
        return new Word(wordIdRef.getKey(), word.getText().toString(), meaning.getText().toString(), new ArrayList<Sentence>() {
            {
                add(s);
            }
        });
    }
}

File: app/src/main/java/unforgettable/azcs/me/unforgettable/AddWordActivity.kt
package unforgettable.azcs.me.unforgettable

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.widget.Button
import android.widget.EditText
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import java.util.*

class AddWordActivity : AppCompatActivity() {
    lateinit var word: EditText
    lateinit var meaning: EditText
    lateinit var sentence: EditText
    lateinit var sentencelist: RecyclerView
    lateinit var addButton: Button
    lateinit var database: FirebaseDatabase
    lateinit var myRef: DatabaseReference
    lateinit var wordIdRef: DatabaseReference
    lateinit var sentenceIdRef: DatabaseReference
    lateinit var user: FirebaseUser

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_word)
        word = findViewById(R.id.word)
        meaning = findViewById(R.id.meaning)
        sentence = findViewById(R.id.sentence)
        sentencelist = findViewById(R.id.sentencesList)
        addButton = findViewById(R.id.addbtn)
        database = FirebaseDatabase.getInstance()
        user = FirebaseAuth.getInstance().currentUser!!
        addButton.setOnClickListener {
            if (Utils.isValid(word) && Utils.isValid(meaning)) {
                myRef = database.getReference(user!!.uid)
                wordIdRef = myRef.push()
                myRef.child(wordIdRef.key).setValue(getWord())
                startActivity(Intent(this@AddWordActivity, MainActivity::class.java))
            }
        }


    }

    private fun getWord(): Word {

        if (TextUtils.isEmpty(sentence.text.toString()))
            return Word(wordIdRef.key, word.text.toString(), meaning.text.toString(), ArrayList())

        sentenceIdRef = wordIdRef.push()
        val s = Sentence(sentence.text.toString(), sentenceIdRef.key)
        return Word(wordIdRef.key, word.text.toString(), meaning.text.toString(), object : ArrayList<Sentence>() {
            init {
                add(s)
            }
        })
    }
}
